std/utf.d

```diff
@@ -3348,85 +3349,79 @@ enum dchar replacementDchar = '\uFFFD';
  *      $(REF byGrapheme, std,uni).
  */
 auto byCodeUnit(R)(R r)
-if (isAutodecodableString!R)
+if (isAutodecodableString!R ||
+    isInputRange!R && isSomeChar!(ElementEncodingType!R) ||
+    (is(R == enum) && is(R : const dchar[])))
 {
-    static struct ByCodeUnitImpl
+    static if (isNarrowString!R ||
+               // This would be cleaner if we had a way to check whether a type
+               // was a range without any implicit conversions.
+               (isAutodecodableString!R && !__traits(hasMember, R, "empty") &&
+                !__traits(hasMember, R, "front") && !__traits(hasMember, R, "popFront")))
     {
-    pure nothrow @nogc:
-
-        @property bool empty() const         { return r.length == 0; }
-        @property auto ref front() inout     { return r[0]; }
-        void popFront()                      { r = r[1 .. $]; }
-        auto ref opIndex(size_t index) inout { return r[index]; }
-
-        @property auto ref back() inout
+        static struct ByCodeUnitImpl
         {
-            return r[$ - 1];
-        }
+        @safe pure nothrow @nogc:
 
-        void popBack()
-        {
-            r = r[0 .. $-1];
-        }
+            @property bool empty() const     { return str.length == 0; }
+            @property auto ref front() inout { return str[0]; }
+            void popFront()                  { str = str[1 .. $]; }
 
-        static if (!isAggregateType!R)
-        {
-            auto opSlice(size_t lower, size_t upper)
-            {
-                return ByCodeUnitImpl(r[lower .. upper]);
-            }
-        }
+            @property auto save() { return ByCodeUnitImpl(str.save); }
 
-        @property size_t length() const
-        {
-            return r.length;
-        }
-        alias opDollar = length;
+            @property auto ref back() inout { return str[$ - 1]; }
+            void popBack()                  { str = str[0 .. $-1]; }
 
-        static if (!isAggregateType!R)
-        {
-            @property auto save()
-            {
-                return ByCodeUnitImpl(r.save);
-            }
-        }
+            auto ref opIndex(size_t index) inout     { return str[index]; }
+            auto opSlice(size_t lower, size_t upper) { return ByCodeUnitImpl(str[lower..upper]); }
 
-      private:
-        R r;
-    }
+            @property size_t length() const { return str.length; }
+            alias opDollar = length;
 
-    static assert(isAggregateType!R || isRandomAccessRange!ByCodeUnitImpl);
+          private:
+            StringTypeOf!R str;
+        }
 
-    return ByCodeUnitImpl(r);
-}
+        static assert(isRandomAccessRange!ByCodeUnitImpl);
 
-/// Ditto
-auto ref byCodeUnit(R)(R r)
-if (!isAutodecodableString!R && isInputRange!R && isSomeChar!(ElementEncodingType!R))
-{
-    // byCodeUnit for ranges and dchar[] is a no-op
-    return r;
+        return ByCodeUnitImpl(r);
+    }
+    else static if (is(R == enum) || is(R : const dchar[]))
```

Comment:
Maybe reading it as a diff is making it hard to see what's actually there? In this case, I'd suggest reading the resulting code rather than trying to parse the diff. It should be much easier to understand that way.